trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/collections/ImmutableMultiMap.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.s3.proxy.server.collections;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ImmutableMultiMap<V>
+        extends MultiMap<V>
+{
+    private ImmutableMultiMap(Map<String, List<V>> items, boolean caseSensitiveKeys)
+    {
+        super(items, caseSensitiveKeys);
+    }
+
+    public static <V> Builder<V> builder(boolean caseSensitiveKeys)
+    {
+        return new Builder<>(new HashMap<>(), caseSensitiveKeys);
+    }
+
+    @Override
+    ImmutableList<V> newCollectionInstance()
+    {
+        return ImmutableList.of();
+    }
+
+    @Override
+    ImmutableList<V> newCollectionInstance(Collection<V> items)
+    {
+        return ImmutableList.copyOf(items);
+    }
+
+    public static <V> ImmutableMultiMap<V> copyOf(Set<Map.Entry<String, List<V>>> entrySet, boolean caseSensitiveKeys)
```

Comment:
For me, the `caseSensitiveKeys` doesn't add much. I'd add a separate method. Something like `copyOfCaseSensitiveKeys`